clients/client-python/gravitino/filesystem/gvfs.py

```diff
@@ -0,0 +1,454 @@
+                    identifier, fileset, fs, virtual_path
+                )
+                self.cache[identifier] = (fileset, fs)
+                context = FilesetContext(identifier, fileset, fs, actual_path)
+                return context
+            raise ValueError(
+                f"Storage under the fileset: `{identifier}` doesn't support now."
+            )
+        finally:
+            write_lock.release()
+
+    def _extract_identifier(self, path) -> NameIdentifier:
+        if path is None or len(path) == 0:
+            raise ValueError("path which need be extracted cannot be null or empty.")
+        match = self._identifier_pattern.match(path)
+        if match and len(match.groups()) == 3:
+            return NameIdentifier.of_fileset(
+                self.metalake,
+                match.group(1),
+                match.group(2),
+                match.group(3),
+            )
+        raise ValueError(f"path: `{path}` doesn't contains valid identifier.")
+
+    def _load_fileset_from_server(self, identifier: NameIdentifier) -> Fileset:
+        catalog: Catalog = self.client.load_catalog(
+            NameIdentifier.of_catalog(
+                identifier.namespace().level(0), identifier.namespace().level(1)
+            )
+        )
+        return catalog.as_fileset_catalog().load_fileset(identifier)
+
+    def _get_actual_path_by_ident(
+        self,
+        identifier: NameIdentifier,
+        fileset: Fileset,
+        fs: FileSystem,
+        virtual_path: str,
+    ) -> str:
+        with_scheme = virtual_path.startswith(self._gvfs_prefix)
+        virtual_location = self._get_virtual_location(identifier, with_scheme)
+        storage_location = fileset.storage_location()
+        if self._check_mount_single_file(fileset, fs):
+            if virtual_path != virtual_location:
+                raise ValueError(
+                    f"Path: {virtual_path} should be same with the virtual location: {virtual_location}"
+                    " when the fileset only mounts a single file."
+                )
+            return storage_location
+        return virtual_path.replace(virtual_location, storage_location, 1)
+
+    def _get_virtual_location(
+        self, identifier: NameIdentifier, with_scheme: bool
+    ) -> str:
+        prefix = self._gvfs_prefix if with_scheme is True else ""
+        return (
+            f"{prefix}"
+            f"/{identifier.namespace().level(1)}"
+            f"/{identifier.namespace().level(2)}"
+            f"/{identifier.name()}"
+        )
+
+    def _check_mount_single_file(self, fileset: Fileset, fs: FileSystem) -> bool:
+        [info] = fs.get_file_info([(self._strip_protocol(fileset.storage_location()))])
+        return info.type is FileType.File
+
+
+@mirror_from(
+    "stream",
+    [
+        "read",
+        "seek",
+        "tell",
+        "write",
+        "readable",
+        "writable",
+        "close",
+        "size",
+        "seekable",
+    ],
```

Comment:
@xloya  what is the meaning of `mirror_from`, does it mean that these APIs will directly offload to the underlying FS?

Reply:
https://github.com/fsspec/filesystem_spec/blob/db6372a4a401b9fd21b35f1e5356c31aed96c98e/fsspec/utils.py#L506